pexconn.go

```diff
@@ -0,0 +1,116 @@
+package torrent
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/anacrolix/log"
+
+	pp "github.com/anacrolix/torrent/peer_protocol"
+)
+
+const (
+	pexRetryDelay = 10 * time.Second
+	pexInterval   = 1 * time.Minute
+)
+
+// per-connection PEX state
+type pexConnState struct {
+	enabled bool
+	xid     pp.ExtensionNumber
+	seq     int
+	timer   *time.Timer
+	gate    chan struct{}
```

Comment:
Anyway to combine this with the timer, or is it not worth it? I had a look and it seems AfterFunc doesn't use Timer.C, and your way saves a goroutine, so is that preferable?

Reply:
That had crossed my mind, too: it was one of the things I tried. It comes out, however, that one can't use Timer.C with AfterFunc, getting a deadlock: apparently, AfterFunc consumes the channel internally. Introduction of this gating channel had helped to resolve the data races flagged by the race detector, lock-less.